contrib/jemalloc/include/jemalloc/internal/jemalloc_internal_inlines_c.h

```diff
@@ -21,6 +21,63 @@
 #include "jemalloc/internal/sz.h"
 #include "jemalloc/internal/witness.h"
 
+JEMALLOC_ALWAYS_INLINE void *
+get_underlying_allocation(tsdn_t *tsdn, void *ptr) {
+	void *ubptr;
+
+#ifndef __CHERI_PURE_CAPABILITY__
+	ubptr = ptr;
+#else
+	rtree_ctx_t *rtree_ctx;
+	rtree_ctx_t rtree_ctx_fallback;
+	extent_t *extent;
+	szind_t szind;
+
+	/*
+	 * These checks catch attacks from an adversary that can manipulate the 
+	 * offset and the bounds of a ptr passed to free() or realloc().
+	 * Reject zero-length capabilities: region arithmetic can misclassify an
+	 * end-of-region pointer as the start of the next region.
+	 */
+	if (unlikely(!cheri_gettag(ptr))) {
+		return NULL;
+	}
+	if (unlikely(cheri_getoffset(ptr) > cheri_getlen(ptr))) {
+		return NULL;
+	}
+	if (unlikely(cheri_getlen(ptr) == 0)) {
+		return NULL;
+	}
+	if (unlikely(cheri_getsealed(ptr))) {
+		return NULL;
+	}
+
+	rtree_ctx = tsdn_rtree_ctx(tsdn, &rtree_ctx_fallback);
+	if (rtree_extent_szind_read(tsdn, &extents_rtree, rtree_ctx,
+	    (uintptr_t)ptr, false, &extent, &szind)) {
+		return NULL;
+	}
+
+	assert(extent_state_get(extent) == extent_state_active);
+	/* Only slab members should be looked up via interior pointers. */
+	assert(extent_addr_get(extent) == ptr || extent_slab_get(extent));
+	assert(szind != SC_NSIZES);
+
+	size_t underlying_size = sz_index2size(szind);
+	void *extent_base = extent->e_addr;
+
+	size_t offset = (uintptr_t)ptr - (uintptr_t)extent_base;
+	size_t region_index = offset / underlying_size;
+	void *region_base = (void *)((uintptr_t)extent_base +
+	    region_index * underlying_size);
+
+	ubptr = cheri_setbounds(
+	    cheri_setaddress(extent_base, (ptraddr_t)region_base),
+	    underlying_size);
+#endif
+	return (ubptr);
```

Comment:
```suggestion
	return ubptr;
```
jemalloc's style does not include the parentheses